cunumeric/vectorize.py

```diff
@@ -0,0 +1,638 @@
+        """Using the magic method __doc__, we KNOW the size of the docstring.
+        We then, just subtract this from the total length of the function
+        """
+        lines_to_skip = 0
+        if func.__doc__ is not None and len(func.__doc__.split("\n")) > 0:
+            lines_to_skip = len(func.__doc__.split("\n"))
+
+        lines = inspect.getsourcelines(func)[0]  # type ignore
+
+        return_lines = []
+        for i in range(lines_to_skip + 1, len(lines)):
+            return_lines.append(lines[i].rstrip())
+        return return_lines
+
+    def _get_return_arguments(self) -> List[str]:
+        """
+        Returns the list of names for return arrays/values
+        """
+        self._func_body = self._get_func_body(self._pyfunc)
+        return_names = []
+        for ln in self._func_body:
+            if "return" in ln:
+                ln = ln.replace("return", "")
+                ln = ln.replace(" ", "")
+                return_names += ln.split(",")
+        # we check if return statement has any special characters since
+        # we don't support cases like "return a+b"
+        for n in return_names:
+            regex = re.compile("[^A-Za-z0-9]")
+            res = regex.findall(n)
+            if len(res) > 0:
+                raise NotImplementedError(
+                    " CuNumeric doesn't support special "
+                    "characters in the return statement of the "
+                    "user-defined function "
+                )
+        return return_names
+
+    def _replace_name(
+        self, name: str, _LOOP_VAR: str, is_gpu: bool = False
+    ) -> str:
+        """
+        add indices to the names of input/output arrays in the function body
+        """
+        if (name in self._arg_names) or (name in self._return_names):
+            return f"{name}[int({_LOOP_VAR})]"
+        else:
+            if is_gpu or ((not is_gpu) and not (name in self._scalar_names)):
+                return f"{name}"
+            else:
+                return f"{name}[0]"
+
+    def _build_gpu_function(self) -> Any:
+        funcid = f"vectorized_{self._pyfunc.__name__}"
+
+        # Preamble
+        lines = ["from numba import cuda"]
+        # we add math and numpy so user-defined functions can use them
+        lines.append("import math")
+        lines.append("import numpy")
+
+        # Signature
+        args = (
+            self._return_names
+            + self._arg_names
+            + self._scalar_names
+            + [_SIZE_VAR]
+            + [_DIM_VAR]
+            + [_PITCHES_VAR]
+            + [_STRIDES_VAR]
+        )
+
+        lines.append("def {}({}):".format(funcid, ",".join(args)))
+        # Initialize the index variable and return immediately
+        # when it exceeds the data size
+        # we compute index for sparse data access when using Legion's
+        # pointer.
+        # a[x][y][z]=a[x*strides[0] + y*strides[1] + z*strides[2]]
+        loop_lines = f"""    local_i = cuda.grid(1)
```

Comment:
I believe this would be equivalent, and keep the block of code together as a block visually
```suggestion
        loop_lines = f"""\
    local_i = cuda.grid(1)
```